File: drivers/go/helpers_test.go
package reddb

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
	"testing"
)

// fakeQuerier records each SQL call and replays scripted JSON responses.
type fakeQuerier struct {
	calls   []fakeCall
	replies [][]byte
	errs    []error
}

type fakeCall struct {
	sql    string
	params []any
}

func (f *fakeQuerier) Query(_ context.Context, sql string, params ...any) ([]byte, error) {
	f.calls = append(f.calls, fakeCall{sql: sql, params: params})
	var body []byte
	if len(f.replies) > 0 {
		body, f.replies = f.replies[0], f.replies[1:]
	}
	var err error
	if len(f.errs) > 0 {
		err, f.errs = f.errs[0], f.errs[1:]
	}
	return body, err
}

func reply(t *testing.T, v any) []byte {
	t.Helper()
	bs, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("marshal reply: %v", err)
	}
	return bs
}

// ----------------------------------------------------------------- KV path

func TestKVPath_QuotesNamespacedKeys(t *testing.T) {
	p, err := KVPath("kv_default", "corpus:version")
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if p != "kv_default.'corpus:version'" {
		t.Fatalf("got %q", p)
	}
}

func TestKVPath_PreservesDotsAndSlashes(t *testing.T) {
	p, _ := KVPath("kv_default", "a/b.c")
	if p != "kv_default.'a/b.c'" {
		t.Fatalf("got %q", p)
	}
}

func TestKVPath_RejectsBadCollection(t *testing.T) {
	_, err := KVPath("bad-name!", "k")
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT, got %v", err)
	}
}

func TestKVValueLiteral(t *testing.T) {
	cases := []struct {
		name string
		in   any
		want string
	}{
		{"nil", nil, "NULL"},
		{"true", true, "true"},
		{"false", false, "false"},
		{"int", int64(42), "42"},
		{"text", "hi", "'hi'"},
		{"escape", "o'reilly", "'o''reilly'"},
		{"object", map[string]any{"a": 1}, `'{"a":1}'`},
	}
	for _, c := range cases {
		got, err := kvValueLiteral(c.in)
		if err != nil {
			t.Fatalf("%s: %v", c.name, err)
		}
		if got != c.want {
			t.Fatalf("%s: got %q want %q", c.name, got, c.want)
		}
	}
}

// ----------------------------------------------------------------- KV ops

func TestKV_Set_EmitsExactKeyPath(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{})}}
	h := NewHelpers(fq).KV()
	if err := h.Set(context.Background(), "characters:hansel", "ok"); err != nil {
		t.Fatalf("set: %v", err)
	}
	sql := fq.calls[0].sql
	if !strings.Contains(sql, "kv_default.'characters:hansel'") {
		t.Fatalf("missing exact key: %s", sql)
	}
	if !strings.Contains(sql, "= 'ok'") {
		t.Fatalf("missing value: %s", sql)
	}
}

func TestKV_Get_ReturnsValueOrNil(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{"rows": []any{map[string]any{"value": "v"}}}),
		reply(t, map[string]any{"rows": []any{}}),
	}}
	h := NewHelpers(fq).KV()
	got, err := h.Get(context.Background(), "k")
	if err != nil || got != "v" {
		t.Fatalf("got %v %v", got, err)
	}
	got, err = h.Get(context.Background(), "k2")
	if err != nil || got != nil {
		t.Fatalf("got %v %v", got, err)
	}
}

func TestKV_Exists_UsesGet(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{"rows": []any{map[string]any{"value": "v"}}}),
		reply(t, map[string]any{"rows": []any{}}),
	}}
	h := NewHelpers(fq).KV()
	r, _ := h.Exists(context.Background(), "k")
	if !r.Exists {
		t.Fatal("want exists=true")
	}
	r, _ = h.Exists(context.Background(), "k2")
	if r.Exists {
		t.Fatal("want exists=false")
	}
}

func TestKV_List_FiltersByPrefixWithoutRewriting(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{"rows": []any{
		map[string]any{"key": "a:1", "value": 1},
		map[string]any{"key": "b:1", "value": 2},
		map[string]any{"key": "a:2", "value": 3},
	}})}}
	h := NewHelpers(fq).KV()
	out, err := h.List(context.Background(), KVListOptions{Prefix: "a:"})
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(out.Items) != 2 {
		t.Fatalf("want 2 rows, got %d", len(out.Items))
	}
	if out.Items[0]["key"] != "a:1" || out.Items[1]["key"] != "a:2" {
		t.Fatalf("rows: %+v", out.Items)
	}
}

func TestKV_List_RejectsNegativeLimit(t *testing.T) {
	h := NewHelpers(&fakeQuerier{}).KV()
	_, err := h.List(context.Background(), KVListOptions{Limit: -1})
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT, got %v", err)
	}
}

// ----------------------------------------------------------------- Queue

func TestQueue_Push_EmitsPriorityAndPayload(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{"affected": 1})}}
	q := NewHelpers(fq).Queue()
	p := 5
	if _, err := q.Push(context.Background(), "jobs", map[string]any{"id": 1}, PushOptions{Priority: &p}); err != nil {
		t.Fatalf("push: %v", err)
	}
	sql := fq.calls[0].sql
	if !strings.HasPrefix(sql, "QUEUE PUSH jobs ") {
		t.Fatalf("prefix: %s", sql)
	}
	if !strings.Contains(sql, "PRIORITY 5") {
		t.Fatalf("priority: %s", sql)
	}
	if !strings.Contains(sql, `{"id":1}`) {
		t.Fatalf("payload: %s", sql)
	}
}

func TestQueue_Len_ReturnsInt(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{"rows": []any{map[string]any{"len": 3}}})}}
	q := NewHelpers(fq).Queue()
	n, err := q.Len(context.Background(), "jobs")
	if err != nil || n != 3 {
		t.Fatalf("len: %d %v", n, err)
	}
}

func TestQueue_Pop_ReturnsPayloads(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{"rows": []any{
		map[string]any{"payload": "a"},
		map[string]any{"payload": "b"},
	}})}}
	q := NewHelpers(fq).Queue()
	out, err := q.Pop(context.Background(), "jobs", 2)
	if err != nil {
		t.Fatalf("pop: %v", err)
	}
	if len(out) != 2 || out[0] != "a" || out[1] != "b" {
		t.Fatalf("payloads: %v", out)
	}
}

func TestQueue_Pop_RejectsNegativeCount(t *testing.T) {
	q := NewHelpers(&fakeQuerier{}).Queue()
	_, err := q.Pop(context.Background(), "jobs", -1)
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT: %v", err)
	}
}

func TestQueue_Push_RejectsInvalidIdentifier(t *testing.T) {
	q := NewHelpers(&fakeQuerier{}).Queue()
	_, err := q.Push(context.Background(), "bad-name!", "x")
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT: %v", err)
	}
}

// ----------------------------------------------------------------- Documents

func TestDocuments_Insert_ReturnsRIDEnvelope(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{"rows": []any{}, "affected": 0}),
		reply(t, map[string]any{
			"rows":     []any{map[string]any{"rid": "doc-1", "body": map[string]any{"name": "alice"}}},
			"affected": 1,
		}),
	}}
	d := NewHelpers(fq).Documents()
	out, err := d.Insert(context.Background(), "people", map[string]any{"name": "alice"})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	if out.Affected != 1 || out.RID != "doc-1" {
		t.Fatalf("envelope: %+v", out)
	}
	if out.Item["rid"] != "doc-1" {
		t.Fatalf("item: %+v", out.Item)
	}
}

func TestDocuments_Get_RaisesNotFoundOnMissing(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{"rows": []any{}})}}
	d := NewHelpers(fq).Documents()
	_, err := d.Get(context.Background(), "people", "doc-1")
	if !IsCode(err, CodeNotFound) {
		t.Fatalf("want NOT_FOUND: %v", err)
	}
}

func TestDocuments_Patch_RejectsJSONPointerPaths(t *testing.T) {
	d := NewHelpers(&fakeQuerier{}).Documents()
	_, err := d.Patch(context.Background(), "people", "doc-1", map[string]any{"a/b": 1})
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT: %v", err)
	}
}

func TestDocuments_List_OrdersByRIDByDefault(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{"rows": []any{
		map[string]any{"rid": "a"},
		map[string]any{"rid": "b"},
	}})}}
	d := NewHelpers(fq).Documents()
	out, err := d.List(context.Background(), "people", ListOptions{})
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(out.Items) != 2 {
		t.Fatalf("rows: %+v", out.Items)
	}
	sql := fq.calls[0].sql
	if !strings.Contains(sql, "ORDER BY rid ASC") {
		t.Fatalf("ordering: %s", sql)
	}
}

func TestDocuments_Insert_PassesThroughExistingCollection(t *testing.T) {
	fq := &fakeQuerier{
		replies: [][]byte{
			reply(t, map[string]any{}),
			reply(t, map[string]any{"rows": []any{map[string]any{"rid": "x"}}, "affected": 1}),
		},
		errs: []error{errors.New("collection already exists"), nil},
	}
	d := NewHelpers(fq).Documents()
	if _, err := d.Insert(context.Background(), "people", map[string]any{"a": 1}); err != nil {
		t.Fatalf("insert: %v", err)
	}
}

// ----------------------------------------------------------------- Tx

func TestTx_Begin_Commit_Rollback_EmitsSQL(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{}),
		reply(t, map[string]any{}),
		reply(t, map[string]any{}),
	}}
	tx := NewHelpers(fq).Tx()
	ctx := context.Background()
	if err := tx.Begin(ctx); err != nil {
		t.Fatalf("begin: %v", err)
	}
	if err := tx.Commit(ctx); err != nil {
		t.Fatalf("commit: %v", err)
	}
	if err := tx.Rollback(ctx); err != nil {
		t.Fatalf("rollback: %v", err)
	}
	want := []string{"BEGIN", "COMMIT", "ROLLBACK"}
	for i, c := range fq.calls {
		if c.sql != want[i] {
			t.Fatalf("call %d: got %q want %q", i, c.sql, want[i])
		}
	}
}

func TestTx_Run_CommitsOnSuccess(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{}), // BEGIN
		reply(t, map[string]any{}), // body work
		reply(t, map[string]any{}), // COMMIT
	}}
	tx := NewHelpers(fq).Tx()
	err := tx.Run(context.Background(), func(child *TxClient) error {
		_, err := child.q.Query(context.Background(), "INSERT INTO t VALUES (1)")
		return err
	})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if fq.calls[0].sql != "BEGIN" || fq.calls[2].sql != "COMMIT" {
		t.Fatalf("expected BEGIN..COMMIT, got %+v", fq.calls)
	}
}

func TestTx_Run_RollsBackOnError(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{}), // BEGIN
		reply(t, map[string]any{}), // ROLLBACK
	}}
	tx := NewHelpers(fq).Tx()
	err := tx.Run(context.Background(), func(_ *TxClient) error {
		return errors.New("boom")
	})
	if err == nil || err.Error() != "boom" {
		t.Fatalf("expected boom, got %v", err)
	}
	if fq.calls[len(fq.calls)-1].sql != "ROLLBACK" {
		t.Fatalf("expected ROLLBACK last, got %+v", fq.calls)
	}
}

// ----------------------------------------------------------------- Documents extra

func TestDocuments_Patch_EmptyRejects(t *testing.T) {
	d := NewHelpers(&fakeQuerier{}).Documents()
	_, err := d.Patch(context.Background(), "people", "doc-1", map[string]any{})
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT, got %v", err)
	}
}

func TestDocuments_Delete_PopulatesDeletedFlag(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{
		reply(t, map[string]any{"affected": 1}),
		reply(t, map[string]any{"affected": 0}),
	}}
	d := NewHelpers(fq).Documents()
	r, _ := d.Delete(context.Background(), "people", "rid-1")
	if !r.Deleted || r.Affected != 1 {
		t.Fatalf("hit: %+v", r)
	}
	r, _ = d.Delete(context.Background(), "people", "rid-missing")
	if r.Deleted || r.Affected != 0 {
		t.Fatalf("miss: %+v", r)
	}
}

// ----------------------------------------------------------------- Queues extra

func TestQueue_Create_EmitsIfNotExists(t *testing.T) {
	fq := &fakeQuerier{replies: [][]byte{reply(t, map[string]any{})}}
	q := NewHelpers(fq).Queues()
	if err := q.Create(context.Background(), "jobs"); err != nil {
		t.Fatalf("create: %v", err)
	}
	if fq.calls[0].sql != "CREATE QUEUE IF NOT EXISTS jobs" {
		t.Fatalf("sql: %q", fq.calls[0].sql)
	}
}

func TestQueue_Create_RejectsBadIdentifier(t *testing.T) {
	q := NewHelpers(&fakeQuerier{}).Queues()
	if err := q.Create(context.Background(), "bad-name!"); !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT, got %v", err)
	}
}

// ---------------------------------------------------------- spec constant

func TestHelperSpecVersion(t *testing.T) {
	if HelperSpecVersion != "1.0" {
		t.Fatalf("HelperSpecVersion: got %q want %q", HelperSpecVersion, "1.0")
	}
}

// ---------------------------------------------------------- decode helpers

func TestAffectedFromBody_HandlesNestedResult(t *testing.T) {
	body := reply(t, map[string]any{"result": map[string]any{"affected": 7}})
	if affectedFromBody(body) != 7 {
		t.Fatalf("nested affected")
	}
}

File: drivers/php/tests/Helpers/HelpersTest.php
<?php
/**
 * Conformance tests mirroring drivers/go/helpers_test.go. They run against
 * a FakeQuerier that records SQL and replays scripted JSON replies — no
 * server needed.
 */

declare(strict_types=1);

namespace Reddb\Tests\Helpers;

use PHPUnit\Framework\TestCase;
use Reddb\Helpers\Documents;
use Reddb\Helpers\Helpers;
use Reddb\Helpers\InvalidArgument;
use Reddb\Helpers\NotFound;
use Reddb\Helpers\Querier;
use Reddb\Helpers\Sql;

final class FakeQuerier implements Querier
{
    /** @var list<array{sql:string,params:array<int,mixed>}> */
    public array $calls = [];
    /** @var list<string> */
    public array $replies = [];
    /** @var list<\Throwable|null> */
    public array $errs = [];

    public function query(string $sql, array $params = []): string
    {
        $this->calls[] = ['sql' => $sql, 'params' => $params];
        $err = array_shift($this->errs);
        $body = array_shift($this->replies) ?? '';
        if ($err !== null && $err !== false) throw $err;
        return $body;
    }

    public function reply(mixed $obj): self
    {
        $this->replies[] = json_encode($obj, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        return $this;
    }

    public function err(\Throwable $e): self
    {
        $this->errs[] = $e;
        return $this;
    }
}

final class HelpersTest extends TestCase
{
    public function testKvPathQuotesNamespacedKeys(): void
    {
        $this->assertSame("kv_default.'corpus:version'", Sql::kvPath('kv_default', 'corpus:version'));
    }

    public function testKvPathPreservesDotsAndSlashes(): void
    {
        $this->assertSame("kv_default.'a/b.c'", Sql::kvPath('kv_default', 'a/b.c'));
    }

    public function testKvPathRejectsBadCollection(): void
    {
        $this->expectException(InvalidArgument::class);
        Sql::kvPath('bad-name!', 'k');
    }

    /** @return iterable<string,array{mixed,string}> */
    public static function kvValueLiteralCases(): iterable
    {
        yield 'null' => [null, 'NULL'];
        yield 'true' => [true, 'true'];
        yield 'false' => [false, 'false'];
        yield 'int' => [42, '42'];
        yield 'text' => ['hi', "'hi'"];
        yield 'escape' => ["o'reilly", "'o''reilly'"];
        yield 'object' => [['a' => 1], "'{\"a\":1}'"];
    }

    /** @dataProvider kvValueLiteralCases */
    public function testKvValueLiteral(mixed $input, string $want): void
    {
        $this->assertSame($want, Sql::kvValueLiteral($input));
    }

    public function testKvSetEmitsExactKeyPath(): void
    {
        $fq = (new FakeQuerier())->reply(new \stdClass());
        (new Helpers($fq))->kv()->set('characters:hansel', 'ok');
        $sql = $fq->calls[0]['sql'];
        $this->assertStringContainsString("kv_default.'characters:hansel'", $sql);
        $this->assertStringContainsString("= 'ok'", $sql);
    }

    public function testKvGetReturnsValueOrNull(): void
    {
        $fq = (new FakeQuerier())
            ->reply(['rows' => [['value' => 'v']]])
            ->reply(['rows' => []]);
        $kv = (new Helpers($fq))->kv();
        $this->assertSame('v', $kv->get('k'));
        $this->assertNull($kv->get('k2'));
    }

    public function testKvExistsUsesGet(): void
    {
        $fq = (new FakeQuerier())
            ->reply(['rows' => [['value' => 'v']]])
            ->reply(['rows' => []]);
        $kv = (new Helpers($fq))->kv();
        $this->assertTrue($kv->exists('k')->exists);
        $this->assertFalse($kv->exists('k2')->exists);
    }

    public function testKvListFiltersByPrefixWithoutRewriting(): void
    {
        $fq = (new FakeQuerier())->reply(['rows' => [
            ['key' => 'a:1', 'value' => 1],
            ['key' => 'b:1', 'value' => 2],
            ['key' => 'a:2', 'value' => 3],
        ]]);
        $out = (new Helpers($fq))->kv()->list(['prefix' => 'a:']);
        $this->assertCount(2, $out->items);
        $this->assertSame('a:1', $out->items[0]['key']);
        $this->assertSame('a:2', $out->items[1]['key']);
    }

    public function testKvListRejectsNegativeLimit(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->kv()->list(['limit' => -1]);
    }

    public function testQueuePushEmitsPriorityAndPayload(): void
    {
        $fq = (new FakeQuerier())->reply(['affected' => 1]);
        (new Helpers($fq))->queue()->push('jobs', ['id' => 1], ['priority' => 5]);
        $sql = $fq->calls[0]['sql'];
        $this->assertStringStartsWith('QUEUE PUSH jobs ', $sql);
        $this->assertStringContainsString('PRIORITY 5', $sql);
        $this->assertStringContainsString('{"id":1}', $sql);
    }

    public function testQueueLenReturnsInt(): void
    {
        $fq = (new FakeQuerier())->reply(['rows' => [['len' => 3]]]);
        $this->assertSame(3, (new Helpers($fq))->queue()->len('jobs'));
    }

    public function testQueuePopReturnsPayloads(): void
    {
        $fq = (new FakeQuerier())->reply(['rows' => [
            ['payload' => 'a'], ['payload' => 'b']]]);
        $out = (new Helpers($fq))->queue()->pop('jobs', 2);
        $this->assertSame(['a', 'b'], $out);
    }

    public function testQueuePopRejectsNegativeCount(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->queue()->pop('jobs', -1);
    }

    public function testQueuePushRejectsInvalidIdentifier(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->queue()->push('bad-name!', 'x');
    }

    public function testDocumentsInsertReturnsRidEnvelope(): void
    {
        $fq = (new FakeQuerier())
            ->reply(['rows' => [], 'affected' => 0])
            ->reply(['rows' => [['rid' => 'doc-1', 'body' => ['name' => 'alice']]],
                     'affected' => 1]);
        $out = (new Helpers($fq))->documents()->insert('people', ['name' => 'alice']);
        $this->assertSame(1, $out->affected);
        $this->assertSame('doc-1', $out->rid);
        $this->assertSame('doc-1', $out->item['rid']);
        // ADR 0067 (#1709): inline JSON literal form — no (body) column list,
        // no quoted-string body coercion.
        $insertSql = '';
        foreach ($fq->calls as $call) {
            if (str_starts_with($call['sql'], 'INSERT')) {
                $insertSql = $call['sql'];
            }
        }
        $this->assertSame(
            'INSERT INTO people DOCUMENT VALUES ({"name":"alice"}) RETURNING *',
            $insertSql
        );
        $this->assertStringNotContainsString('(body)', $insertSql);
        $this->assertStringNotContainsString("('{", $insertSql);
    }

    public function testDocumentsGetRaisesNotFoundOnMissing(): void
    {
        $fq = (new FakeQuerier())->reply(['rows' => []]);
        $this->expectException(NotFound::class);
        (new Helpers($fq))->documents()->get('people', 'doc-1');
    }

    public function testDocumentsPatchRejectsJsonPointerPaths(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->documents()
            ->patch('people', 'doc-1', ['a/b' => 1]);
    }

    public function testDocumentsListOrdersByRidByDefault(): void
    {
        $fq = (new FakeQuerier())->reply(['rows' => [['rid' => 'a'], ['rid' => 'b']]]);
        $out = (new Helpers($fq))->documents()->list('people');
        $this->assertCount(2, $out->items);
        $this->assertStringContainsString('ORDER BY rid ASC', $fq->calls[0]['sql']);
    }

    public function testDocumentsInsertPassesThroughExistingCollection(): void
    {
        $fq = (new FakeQuerier())
            ->err(new \RuntimeException('collection already exists'))
            ->reply(new \stdClass())
            ->reply(['rows' => [['rid' => 'x']], 'affected' => 1]);
        // err array shorter than calls — second call has no err.
        (new Helpers($fq))->documents()->insert('people', ['a' => 1]);
        $this->assertTrue(true); // reached without throwing
    }

    public function testAffectedFromBodyHandlesNestedResult(): void
    {
        $body = (string) json_encode(['result' => ['affected' => 7]]);
        $this->assertSame(7, Sql::affectedFromBody($body));
    }

    // --- SDK Helper Spec v1.0 surface -----------------------------------

    public function testHelperSpecVersionConstant(): void
    {
        $this->assertSame('1.0', Helpers::HELPER_SPEC_VERSION);
        $this->assertSame('1.0', (new Helpers(new FakeQuerier()))->helperSpecVersion());
    }

    public function testQueuesIsPluralAliasOfQueue(): void
    {
        $h = new Helpers(new FakeQuerier());
        $this->assertInstanceOf(\Reddb\Helpers\Queue::class, $h->queues());
    }

    public function testQueueCreateEmitsIfNotExists(): void
    {
        $fq = (new FakeQuerier())->reply(new \stdClass());
        (new Helpers($fq))->queues()->create('jobs');
        $this->assertSame('CREATE QUEUE IF NOT EXISTS jobs', $fq->calls[0]['sql']);
    }

    public function testQueueCreateRejectsBadIdentifier(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->queues()->create('bad-name!');
    }

    public function testKvDeleteReturnsEnvelope(): void
    {
        $fq = (new FakeQuerier())
            ->reply(['affected' => 1])
            ->reply(['affected' => 0]);
        $kv = (new Helpers($fq))->kv();
        $hit = $kv->delete('k');
        $this->assertSame(1, $hit->affected);
        $this->assertTrue($hit->deleted);
        $miss = $kv->delete('k');
        $this->assertSame(0, $miss->affected);
        $this->assertFalse($miss->deleted);
    }

    public function testDocumentsDeleteReturnsEnvelope(): void
    {
        $fq = (new FakeQuerier())->reply(['affected' => 0]);
        $res = (new Helpers($fq))->documents()->delete('people', 'missing');
        $this->assertSame(0, $res->affected);
        $this->assertFalse($res->deleted);
    }

    public function testDocumentsPatchRejectsEmptyPatch(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->documents()->patch('people', 'doc-1', []);
    }

    public function testGenericQueryRejectsEmptySql(): void
    {
        $this->expectException(InvalidArgument::class);
        (new Helpers(new FakeQuerier()))->query('   ');
    }

    public function testBulkInsertEmptyIsNoOp(): void
    {
        $fq = new FakeQuerier();
        $res = (new Helpers($fq))->bulkInsert('people', []);
        $this->assertSame(0, $res->affected);
        $this->assertSame([], $res->rids);
        $this->assertCount(0, $fq->calls);
    }

    public function testBulkInsertPreservesOrder(): void
    {
        $fq = (new FakeQuerier())
            // first insert: ensureCollection + insert
            ->reply(new \stdClass())
            ->reply(['rows' => [['rid' => 'r-0']], 'affected' => 1])
            ->reply(new \stdClass())
            ->reply(['rows' => [['rid' => 'r-1']], 'affected' => 1]);
        $res = (new Helpers($fq))->bulkInsert('people', [['i' => 0], ['i' => 1]]);
        $this->assertSame(2, $res->affected);
        $this->assertSame(['r-0', 'r-1'], $res->rids);
    }

    public function testTxBeginCommitRollbackEmitSql(): void
    {
        $fq = (new FakeQuerier())->reply(new \stdClass())->reply(new \stdClass())->reply(new \stdClass());
        $tx = (new Helpers($fq))->tx();
        $tx->begin();
        $tx->commit();
        $tx->rollback();
        $this->assertSame('BEGIN', $fq->calls[0]['sql']);
        $this->assertSame('COMMIT', $fq->calls[1]['sql']);
        $this->assertSame('ROLLBACK', $fq->calls[2]['sql']);
    }

    public function testTxRunCommitsOnSuccess(): void
    {
        $fq = (new FakeQuerier())->reply(new \stdClass())->reply(new \stdClass());
        (new Helpers($fq))->tx()->run(function ($tx): void {
            // no-op body
        });
        $this->assertSame('BEGIN', $fq->calls[0]['sql']);
        $this->assertSame('COMMIT', $fq->calls[1]['sql']);
    }

    public function testTxRunRollsBackAndRethrowsOnFailure(): void
    {
        $fq = (new FakeQuerier())->reply(new \stdClass())->reply(new \stdClass());
        try {
            (new Helpers($fq))->tx()->run(function ($tx): void {
                throw new \RuntimeException('boom');
            });
            $this->fail('expected exception to propagate');
        } catch (\RuntimeException $e) {
            $this->assertSame('boom', $e->getMessage());
        }
        $this->assertSame('BEGIN', $fq->calls[0]['sql']);
        $this->assertSame('ROLLBACK', $fq->calls[1]['sql']);
    }

    public function testTxRunRejectsNesting(): void
    {
        $fq = (new FakeQuerier())->reply(new \stdClass())->reply(new \stdClass());
        $this->expectException(InvalidArgument::class);
        (new Helpers($fq))->tx()->run(function ($tx): void {
            $tx->run(function ($inner): void {});
        });
    }
}
